http-client/src/client.rs

```diff
@@ -23,6 +22,71 @@ impl HttpClient {
 		let transport = HttpTransportClient::new(target, config).map_err(|e| Error::TransportError(Box::new(e)))?;
 		Ok(Self { transport, request_id: AtomicU64::new(0) })
 	}
+
+	/// Perform a batch request towards the server.
+	///
+	/// Returns `Ok` if all requests were answered successfully.
+	/// Returns `Error` if any of the requests fails.
+	//
+	// TODO(niklasad1): maybe simplify generic `requests`, it's quite unreadable.
+	pub async fn batch_request<'a, T>(
+		&self,
+		requests: impl IntoIterator<Item = (impl Into<String>, impl Into<jsonrpc::Params>)>,
+	) -> Result<Vec<T>, Error>
+	where
+		T: DeserializeOwned,
+	{
+		let mut calls = Vec::new();
+		// NOTE(niklasad1): `ID` is not necessarily monotonically increasing.
+		let mut ids = Vec::new();
+
+		for (method, params) in requests.into_iter() {
+			let id = self.request_id.fetch_add(1, Ordering::SeqCst);
+			calls.push(jsonrpc::Call::MethodCall(jsonrpc::MethodCall {
+				jsonrpc: jsonrpc::Version::V2,
+				method: method.into(),
+				params: params.into(),
+				id: jsonrpc::Id::Num(id),
+			}));
+			ids.push(id);
+		}
+
+		let batch_request = jsonrpc::Request::Batch(calls);
+		let response = self
+			.transport
+			.send_request_and_wait_for_response(batch_request)
+			.await
+			.map_err(|e| Error::TransportError(Box::new(e)))?;
+
+		match response {
+			jsonrpc::Response::Single(_) => {
+				Err(Error::Custom("Server replied with single response to a batch request".to_string()))
+			}
+			jsonrpc::Response::Notif(_notif) => {
+				Err(Error::Custom("Server replied with notification to with a batch request".to_string()))
+			}
+			jsonrpc::Response::Batch(rps) => {
+				// NOTE: `JsonValue` is a placeholder and will be replaced in the loop below.
+				let mut json_responses = vec![jsonrpc::JsonValue::Number(0.into()); ids.len()];
+				for rp in rps {
+					let id = match rp.id().as_number() {
+						Some(n) => *n,
+						_ => return Err(Error::InvalidRequestId),
+					};
+					// NOTE(niklasad1): O(n)
+					let pos = match ids.iter().position(|i| i == &id) {
+						Some(id) => id,
+						None => return Err(Error::InvalidRequestId),
+					};
+					let json_val: jsonrpc::JsonValue = rp.try_into().map_err(Error::Request)?;
+					json_responses[pos] = json_val;
+				}
+				let responses: Result<_, _> =
+					json_responses.into_iter().map(|val| jsonrpc::from_value(val).map_err(Error::ParseError)).collect();
```

Comment:
yeah, I think it did like this because `jsonrpc::from_value` returns `T.

Then we either to add `T: Default` or put in an `Option` where the json_responses is declared.

Reply:
`T: Default` sounds reasonable to me.